dmriprep/utils/vectors.py

```diff
@@ -1,482 +1,346 @@
-#!/usr/bin/env python
-# -*- coding: utf-8 -*-
-import os
-import nibabel as nib
+"""Utilities to operate on diffusion gradients."""
+from pathlib import Path
+from itertools import permutations
+import nibabel as nb
 import numpy as np
-
-
-def make_gradient_table(bvals, bvecs, B0_NORM_EPSILON):
+from dipy.core.gradients import round_bvals
+
+B0_THRESHOLD = 50
+BVEC_NORM_EPSILON = 0.1
+
+
+class DiffusionGradientTable:
+    """Data structure for DWI gradients."""
+
+    __slots__ = ['_affine', '_gradients', '_b_scale', '_bvecs', '_bvals', '_normalized',
+                 '_b0_thres', '_bvec_norm_epsilon']
+
+    def __init__(self, dwi_file=None, bvecs=None, bvals=None, rasb_file=None,
+                 b_scale=True, b0_threshold=B0_THRESHOLD, bvec_norm_epsilon=BVEC_NORM_EPSILON):
+        """
+        Create a new table of diffusion gradients.
+
+        Parameters
+        ----------
+        dwi_file : str or os.pathlike or nibabel.spatialimage
+            File path to the diffusion-weighted image series to which the
+            bvecs/bvals correspond.
+        bvals : str or os.pathlike or numpy.ndarray
+            File path of the b-values.
+        bvecs : str or os.pathlike or numpy.ndarray
+            File path of the b-vectors.
+        rasb_file : str or os.pathlike
+            File path to a RAS-B gradient table. If rasb_file is provided,
+            then bvecs and bvals will be dismissed.
+        b_scale : bool
+            Whether b-values should be normalized.
+
+        """
+        self._b_scale = b_scale
+        self._b0_thres = b0_threshold
+        self._bvec_norm_epsilon = bvec_norm_epsilon
+        self._gradients = None
+        self._bvals = None
+        self._bvecs = None
+        self._affine = None
+        self._normalized = False
+
+        if dwi_file is not None:
+            self.affine = dwi_file
+        if rasb_file is not None:
+            self.gradients = rasb_file
+            self.generate_vecval()
+        elif dwi_file and bvecs and bvals:
+            self.bvecs = bvecs
+            self.bvals = bvals
+            self.generate_rasb()
+
+    @property
+    def affine(self):
+        return self._affine
+
+    @property
+    def gradients(self):
+        """Get gradient table (rasb)."""
+        return self._gradients
+
+    @property
+    def bvecs(self):
+        return self._bvecs
+
+    @property
+    def bvals(self):
+        return self._bvals
+
+    @property
+    def normalized(self):
+        return self._normalized
+
+    @affine.setter
+    def affine(self, value):
```

Comment:
Python setters/getters are cool

Reply:
Ahh, I wasn't sure before whether you had been thinking of setter/getter decorators, but this looks like a better alternative!